### PR TITLE
Refcounted DB columns

### LINKED ISSUE #3169: Receipts, unlike transactions, are not refcoutned.
The same receipt (and the same transaction) can be in multiple chunks. Thus, when we garbage collect chunks, we need to make sure it is the last occurrence of a particular receipt or transaction. For txs we have a refcount in place, but for receipts we do not.

In `clear_chunk_data`:

```
            for receipt in chunk.receipts {
                self.gc_col(ColReceiptIdToShardId, &receipt.receipt_id.into());
            }
            for transaction in chunk.transactions {
                self.gc_col_transaction(transaction.get_hash())?;
            }
```

We need to implement the same refcounting for receipts.

UPD: When fixed, uncomment the debug assert in core/store/src/lib.rs (search for 3169 there)

Reproduced here: http://nayduck.eastus.cloudapp.azure.com:3000/#/test/12420

### PR DESCRIPTION
**Pull request by mikhailOK**
- Move refcount logic for ColState from Trie to Store
- Change refcount for ColState 4 byte to 8 byte
- Use the new logic for ColTransactions, deprecate ColTransactionRefCount
- Use it for ColReceiptIdToShardId

This requires a storage version upgrade.

Fixes #3169

Test plan
---------
TODO

### CODE REVIEW

File: core/store/src/db/refcount.rs [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        return;
+    }
+    let (result_bytes, result_rc) = decode_value_with_rc(result);
+    if result_rc == 0 {
+        result.extend_from_slice(val);
+    } else {
+        let rc = result_rc + add_rc;
+        debug_assert!(result_rc <= 0 || add_rc <= 0 || result_bytes == bytes);
+        match rc.cmp(&0) {
+            Ordering::Less => {
```
Comment (bowenwang1996) [inline core/store/src/db/refcount.rs:33]: How can this case happen?

File: core/store/src/db/v6.rs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    }
+    Some(result)
+}
+
+fn vec_to_rc(bytes: &[u8]) -> i32 {
+    let mut cursor = Cursor::new(&bytes[bytes.len() - 4..]);
+    cursor.read_i32::<LittleEndian>().unwrap()
+}
+
+fn merge_refcounted_records_v6(result: &mut Vec<u8>, val: &[u8]) -> () {
```
Comment (bowenwang1996) [inline core/store/src/db/v6.rs:31]:
```suggestion
fn merge_refcounted_records_v6(result: &mut Vec<u8>, val: &[u8]) {
```

**Comment by bowenwang1996** [comment]
It looks like nightly tests regressed.

**Comment by SkidanovAlex** [comment]
> It looks like nightly tests regressed.

It's a Nayduck issue, we'll need to restart.

File: core/store/src/db/refcount.rs (hunk 1 above)
Comment (mikhailOK) [reply core/store/src/db/refcount.rs:33]: Depends on the order of merging - if it happens in order of insertion intermediate results shouldn't go negative, but easier not to think about it

Comment (mikhailOK) [reply core/store/src/db/refcount.rs:33]: Looks like it happens in order of insertion, so it would be safe to debug_assert here

**Comment by bowenwang1996** [comment]
@SkidanovAlex looks like there is still regression from the nayduck run you linked to.

**Comment by mikhailOK** [comment]
No regression, all failures are broken or flaky tests

**Comment by bowenwang1996** [comment]
I don't think `contracts/gibberish.py` and `sanity/staking_repro2.py` were flaky.

**Comment by mikhailOK** [comment]
both pass
http://nayduck.eastus.cloudapp.azure.com:3000/run/166#/test/17691
http://nayduck.eastus.cloudapp.azure.com:3000/run/166#/test/17729

**Pull request merged**